kernel/src/hil/uart.rs

```diff
@@ -48,20 +53,29 @@ pub trait UART {
     /// called when events finish.
     fn set_client(&self, client: &'static Client);
 
-    /// Initialize UART
+    /// Configure UART
+    ///
+    /// Returns SUCCESS, or
     ///
-    /// Panics if UARTParams are invalid for the current chip.
-    fn init(&self, params: UARTParams);
+    /// - EOFF: The underlying hardware is currently not available, perhaps
+    ///         because it has not been initialized or in the case of a shared
+    ///         hardware USART controller because it is set up for SPI.
+    /// - EINVAL: Impossible parameters (e.g. a `baud_rate` of 0)
+    /// - ENOSUPPORT: The underlying UART cannot satisfy this configuration.
+    fn configure(&self, params: UARTParameters) -> ReturnCode;
 
     /// Transmit data.
-    fn transmit(&self, tx_data: &'static mut [u8], tx_len: usize);
+    fn transmit(&self, tx_data: &'static mut [u8], tx_len: usize) -> ReturnCode;
 
     /// Receive data until buffer is full.
-    fn receive(&self, rx_buffer: &'static mut [u8], rx_len: usize);
-
-    /// Abort any ongoing receive transfers and return what is in the
-    /// receive buffer with the `receive_complete` callback.
-    fn abort_receive(&self);
+    fn receive(&self, rx_buffer: &'static mut [u8], rx_len: usize) -> ReturnCode;
+
+    /// Abort any ongoing receive transfers. The `receieve_callback` will be
+    /// called with `AbortedError`. Callers should take care to note that an
+    /// active receive may complete before this call occurs, in which case the
+    /// `receive_callback` will be called with `CommandComplete` and this method
+    /// will return `EALREADY`.
+    fn abort_receive(&self) -> ReturnCode;
```

Comment:
The PR description doesn't seem to match this comment. This doesn't say whether data will be returned or not.

In either case, returning what has been received so far was the entire purpose of adding this function. Without it, there is no way to implement a UART receiver that can take in variable length messages without doing it byte-by-byte.